proposals/p2188.md

```diff
@@ -0,0 +1,1689 @@
+callee in a function call expression, and an alternative pattern is
+syntactically a function call expression whose argument list contains at least
+one _proper-pattern_.
+
+If a _callee-expression_ is provided, it is required to name a choice type
+alternative that has a parameter list, and the scrutinee is implicitly converted
+to that choice type. Otherwise, the scrutinee is required to be of some choice
+type, and the designator is looked up in that type and is required to name an
+alternative with a parameter list if and only if a _tuple-pattern_ is specified.
+
+The pattern matches if the active alternative in the scrutinee is the specified
+alternative, and the arguments of the alternative match the given tuple pattern
+(if any).
+
+```
+choice Optional(T:! Type) {
+  None,
+  Some(T)
+}
+
+match (Optional(i32).None) {
+  // ✅ `.None` resolved to `Optional(i32).None`.
+  case .None => {}
+  // ✅ `.Some` resolved to `Optional(i32).Some`.
+  case .Some(n: i32) => { Print("{0}", n); }
+  // ❌ Error, no such alternative exists.
+  case .Other => {}
+}
+
+class X {
+  external impl as ImplicitAs(Optional(i32));
+}
+
+match ({} as X) {
+  // ✅ OK, but expression pattern.
+  case Optional(i32).None => {}
+  // ✅ OK, implicitly converts to `Optional(i32)`.
+  case Optional(i32).Some(n: i32) => { Print("{0}", n); }
+}
+```
+
+Note that a pattern of the form `Optional(T).None` is an expression pattern and
+is compared using `==`.
+
+### Templates
+
+Any checking of the type of the scrutinee against the type of the pattern that
+cannot be performed because the type of the scrutinee involves a template
+parameter is deferred until the template parameter's value is known. During
+instantiation, patterns that are not meaningful due to a type error are instead
+treated as not matching. This includes cases where an `==` fails because of a
+missing `EqWith` implementation.
+
+```
+fn TypeName[template T:! Type](x: T) -> String {
+  match (x) {
+    // ✅ OK, the type of `x` is a template parameter.
+    case _: i32 => { return "int"; }
+    case _: bool => { return "bool"; }
+    case _: auto* => { return "pointer"; }
+    default => { return "unknown"; }
+  }
+}
+```
+
+Cases where the match is invalid for reasons not involving the template
+parameter are rejected when type-checking the template:
+
+```
+fn MeaninglessMatch[template T:! Type](x: T*) {
+  match (*x) {
+    // ✅ OK, `T` could be a tuple.
+    case (_: auto, _: auto) => {}
+    default => {}
+  }
+  match (x->y) {
+    // ✅ OK, `T.y` could be a tuple.
+    case (_: auto, _: auto) => {}
+    default => {}
+  }
+  match (x) {
+    // ❌ Error, tuple pattern cannot match value of non-tuple type `T*`.
+    case (_: auto, _: auto) => {}
+    default => {}
+  }
+}
+```
+
+### Guards
+
+We allow `case`s within a `match` statement to have _guards_. These are not part
+of pattern syntax, but instead are specific to `case` syntax:
+
+-   _case_ ::= `case` _pattern_ [`if` _expression_]? `=>` _block_
```

Comment:
Idle and pretty minor comment -- I wonder if we should consider alternative syntaxes for the `=>` part of this... Specifically in case we end up wanting `=>` for any other parts of the grammar.

Not super important though, as this might be the *least* impactful syntax question here, and the easiest for us to change our minds about subsequently. =D

Reply:
This proposal isn't really considering syntax options for `case` / `match` in general; all this section is trying to describe is where the `if` _expression_ part fits into the existing (but not in an approved design) syntax. We should consider alternatives for `match` and `case` syntax, but not in this proposal.